tag_tracer.go

```diff
@@ -0,0 +1,265 @@
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/connmgr"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+)
+
+var (
+	// GossipSubConnTagValueDirectPeer is the connection manager tag value to
+	// apply to direct peers. This should be high, as we want to prioritize these
+	// connections above all others.
+	GossipSubConnTagValueDirectPeer = 1000
+
+	// GossipSubConnTagValueMeshPeer is the connection manager tag value to apply to
+	// peers in a topic mesh. If a peer is in the mesh for multiple topics, their
+	// connection will be tagged separately for each.
+	GossipSubConnTagValueMeshPeer = 20
+
+	// GossipSubConnTagBumpMessageDelivery is the amount to add to the connection manager
+	// tag that tracks message deliveries. Each time a peer is the first to deliver a
+	// message within a topic, we "bump" a tag by this amount, up to a maximum
+	// of GossipSubConnTagMessageDeliveryCap.
+	// Note that the delivery tags decay over time, decreasing by GossipSubConnTagDecayAmount
+	// at every GossipSubConnTagDecayInterval.
+	GossipSubConnTagBumpMessageDelivery = 1
+
+	// GossipSubConnTagDecayInterval is the decay interval for decaying connection manager tags.
+	GossipSubConnTagDecayInterval = 10 * time.Minute
+
+	// GossipSubConnTagDecayAmount is subtracted from decaying tag values at each decay interval.
+	GossipSubConnTagDecayAmount = 1
+
+	// GossipSubConnTagMessageDeliveryCap is the maximum value for the connection manager tags that
+	// track message deliveries.
+	GossipSubConnTagMessageDeliveryCap = 15
+)
+
+// tagTracer is an internal tracer that applies connection manager tags to peer
+// connections based on their behavior.
+//
+// We tag a peer's connections for the following reasons:
+// - Directly connected peers are tagged with GossipSubConnTagValueDirectPeer (default 1000).
+// - Mesh peers are tagged with a value of GossipSubConnTagValueMeshPeer (default 20).
+//   If a peer is in multiple topic meshes, they'll be tagged for each.
+// - For each message that we receive, we bump a delivery tag for peer that delivered the message
+//   first.
+//   The delivery tags have a maximum value, GossipSubConnTagMessageDeliveryCap, and they decay at
+//   a rate of GossipSubConnTagDecayAmount / GossipSubConnTagDecayInterval.
+type tagTracer struct {
+	sync.RWMutex
+
+	cmgr     connmgr.ConnManager
+	msgID    MsgIdFunction
+	decayer  connmgr.Decayer
+	decaying map[string]connmgr.DecayingTag
+	direct   map[peer.ID]struct{}
+
+	// a map of message ids to the set of peers who delivered the message after the first delivery,
+	// but before the message was finished validating
+	nearFirst map[string]map[peer.ID]struct{}
+}
+
+func newTagTracer(cmgr connmgr.ConnManager) *tagTracer {
+	decayer, ok := connmgr.SupportsDecay(cmgr)
+	if !ok {
+		log.Warnf("connection manager does not support decaying tags, delivery tags will not be applied")
+	}
+	return &tagTracer{
+		cmgr:      cmgr,
+		msgID:     DefaultMsgIdFn,
```

Comment:
Do we always want to use the default? shouldn't we use the message ID function that the user has set?

Reply:
We set it to the router's `msgID` func in the `Start` method. Although now that I removed the background process we don't really need a `Start` method and could just put it in the constructor instead.

Reply:
The actual MsgID function is not known when it is constructed, it might be changed by the options.
So keep the Start method :)

Reply:
vyzo reminded me that we should keep it in `Start`, since the message id is set in an option that might be applied after the `tagTracer` is created.